File: solutions/graphs/alternatives/261-graph-valid-tree.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Union-Find data structure for cycle detection
 */
class UnionFind {
    constructor(n) {
        this.parent = Array.from({length: n}, (_, i) => i);
        this.rank = new Array(n).fill(0);
        this.components = n;
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]); // Path compression
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX === rootY) {
            return false; // Cycle detected
        }

        // Union by rank
        if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else if (this.rank[rootX] > this.rank[rootY]) {
            this.parent[rootY] = rootX;
        } else {
            this.parent[rootY] = rootX;
            this.rank[rootX]++;
        }

        this.components--;
        return true;
    }
}

/**
 * Main solution for Problem 261: Graph Valid Tree
 *
 * @param {number} n - Number of nodes (0 to n-1)
 * @param {number[][]} edges - Array of edges [u, v]
 * @return {boolean} - True if the graph forms a valid tree
 *
 * Time Complexity: O(n * α(n)) - Union-Find with path compression
 * Space Complexity: O(n) - parent and rank arrays
 */
function solve(n, edges) {
    // A tree with n nodes must have exactly n-1 edges
    if (edges.length !== n - 1) {
        return false;
    }

    const uf = new UnionFind(n);

    // Process each edge
    for (const [u, v] of edges) {
        // If union returns false, a cycle was detected
        if (!uf.union(u, v)) {
            return false;
        }
    }

    // Check if all nodes are connected (exactly 1 component)
    return uf.components === 1;
}

/**
 * Test cases for Problem 261: Graph Valid Tree
 */
function testSolution() {
    console.log('Testing 261. Graph Valid Tree');

    // Test case 1: Valid tree
    const result1 = solve(5, [[0,1],[0,2],[0,3],[1,4]]);
    console.assert(result1 === true, `Test 1 failed: expected true, got ${result1}`);

    // Test case 2: Has cycle
    const result2 = solve(5, [[0,1],[1,2],[2,3],[1,3],[1,4]]);
    console.assert(result2 === false, `Test 2 failed: expected false, got ${result2}`);

    // Test case 3: Not connected (too few edges)
    const result3 = solve(5, [[0,1],[2,3]]);
    console.assert(result3 === false, `Test 3 failed: expected false, got ${result3}`);

    // Test case 4: Single node
    const result4 = solve(1, []);
    console.assert(result4 === true, `Test 4 failed: expected true, got ${result4}`);

    // Test case 5: Two nodes connected
    const result5 = solve(2, [[0,1]]);
    console.assert(result5 === true, `Test 5 failed: expected true, got ${result5}`);

    // Test case 6: Two nodes not connected
    const result6 = solve(2, []);
    console.assert(result6 === false, `Test 6 failed: expected false, got ${result6}`);

    // Test case 7: Too many edges
    const result7 = solve(3, [[0,1],[1,2],[0,2],[0,1]]);
    console.assert(result7 === false, `Test 7 failed: expected false, got ${result7}`);

    // Test case 8: Linear tree
    const result8 = solve(4, [[0,1],[1,2],[2,3]]);
    console.assert(result8 === true, `Test 8 failed: expected true, got ${result8}`);

    console.log('All test cases passed for 261. Graph Valid Tree!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 261. Graph Valid Tree ===');
    console.log('Category: Graphs');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on graphs concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: docs/solutions/graphs/261-graph-valid-tree.py
"""
# Difficulty: Medium

# 261. Graph Valid Tree

You have a graph of n nodes labeled from 0 to n - 1. You are given an integer n and a list of edges where edges[i] = [ai, bi] indicates that there is an undirected edge between nodes ai and bi in the graph.

Return true if the edges of the given graph make up a valid tree, and false otherwise.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[[0,1]</dd>
<dt>Output:</dt>
<dd>"Expected {expected}, got {result}"</dd>
<dt>Explanation:</dt>
<dd>Graph edges form a valid tree if connected and has n-1 edges</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
A valid tree has exactly n-1 edges and is fully connected with no cycles. Use Union-Find to detect cycles and count components. Valid if no cycles and all nodes in one component.

### APPROACH:
1. **Check edge count**: If edges != n-1, return False
2. **Build adjacency list**: Create graph as defaultdict(list), add edges in both directions
3. **Initialize visited set**: Create empty set to track visited nodes
4. **Define DFS function**: Implement dfs(node, parent) to traverse graph
5. **Check for cycles**: In DFS, if neighbor is visited and not parent, cycle detected
6. **Mark as visited**: Add current node to visited set, recurse on unvisited neighbors
7. **Check connectivity**: After DFS from node 0, verify len(visited) == n
8. **Return result**: Return True if no cycles and all nodes visited, else False

### WHY THIS WORKS:
- Tree properties: n nodes, n-1 edges, all nodes connected, no cycles
- Check edge count == n-1 (necessary but not sufficient)
- DFS/BFS from any node: all nodes should be visited (connected)
- Union-find alternative: no cycles means each edge connects different components
- O(n + e) time: DFS/BFS, O(n) space for visited array

### EXAMPLE WALKTHROUGH:
```
Input: n = 5, edges = [[0,1],[0,2],[0,3],[1,4]]
Step 1: Check edge count
  n-1 = 4 edges (necessary for tree) ✓

Step 2: Build adjacency list
  {0: [1,2,3], 1: [0,4], 2: [0], 3: [0], 4: [1]}

Step 3: DFS to check connectivity
  Visit 0 → mark visited
  Visit 1 → mark visited
  Visit 4 → mark visited
  Visit 2 → mark visited
  Visit 3 → mark visited

Step 4: Verify all nodes visited
  visited = {0,1,2,3,4}, count = 5 = n ✓

Output: True (forms valid tree)
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

from collections import defaultdict, deque

from typing import List, Optional, Dict, Tuple

class Solution:
    """
    Solution class for determining if a graph is a valid tree.
    A valid tree must:
    1. Be fully connected (all nodes can be reached)
    2. Have no cycles
    3. Have n-1 edges for n nodes
    """
    
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        """
        Determines if the given graph represents a valid tree.
        
        Args:
            n: Number of nodes (labeled from 0 to n-1)
            edges: List of edges, where each edge is [node1, node2]
            
        Returns:
            bool: True if the graph is a valid tree, False otherwise
        """
        # First check: A tree with n nodes must have exactly n-1 edges
        if len(edges) != n - 1:
            return False
            
        # Create adjacency list representation of the graph
        adj_list = defaultdict(list)
        for u, v in edges:
            adj_list[u].append(v)
            adj_list[v].append(u)
            
        # Use BFS to check if the graph is fully connected and has no cycles
        visited = set()
        queue = deque([0])  # Start from node 0
        
        while queue:
            node = queue.popleft()
            visited.add(node)
            
            for neighbor in adj_list[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        
        # If we visited all nodes, and we already verified we have n-1 edges,
        # then this must be a valid tree
        return len(visited) == n

    def validTree_dfs(self, n: int, edges: List[List[int]]) -> bool:
        """
        Alternative implementation using DFS.
        
        Args:
            n: Number of nodes (labeled from 0 to n-1)
            edges: List of edges, where each edge is [node1, node2]
            
        Returns:
            bool: True if the graph is a valid tree, False otherwise
        """
        if len(edges) != n - 1:
            return False
            
        # Create adjacency list
        adj_list = defaultdict(list)
        for u, v in edges:
            adj_list[u].append(v)
            adj_list[v].append(u)
            
        visited = set()
        
        def dfs(node: int, parent: int) -> bool:
            visited.add(node)
            
            for neighbor in adj_list[node]:
                if neighbor == parent:
                    continue
                if neighbor in visited:
                    return False
                if not dfs(neighbor, node):
                    return False
            return True
            
        return dfs(0, -1) and len(visited) == n

def test_solution():
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Valid tree
    result = solution.validTree(5, [[0,1],[0,2],[0,3],[1,4]])
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Has cycle (not a tree)
    result = solution.validTree(5, [[0,1],[1,2],[2,3],[1,3],[1,4]])
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: Single node (valid tree)
    result = solution.validTree(1, [])
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 4: Disconnected graph (not a tree)
    result = solution.validTree(4, [[0,1],[2,3]])
    expected = False
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 5: Using DFS version
    result = solution.validTree_dfs(5, [[0,1],[0,2],[0,3],[1,4]])
    expected = True
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 261. Graph Valid Tree")
